### PR TITLE
[aten][autograd] Add narrow_scatter operator with autograd and torch.compile support

### LINKED ISSUE #177372: Add torch.narrow_scatter operator to complete the *_copy/*_scatter operator family.
### The feature, motivation and pitch

PyTorch currently has `narrow_copy` but lacks the corresponding `narrow_scatter` operator, breaking the pattern where every `*_copy` operation has a matching `*_scatter` operation:
- `slice_copy` / `slice_scatter` - exists
- `select_copy` / `select_scatter` - exists
- `narrow_copy` / `narrow_scatter` - missing

Without `narrow_scatter`, the functorch functionalization layer can't properly handle mutations through narrow views. Right now users have to use mutation-based workarounds that don't work with functorch:

```python
# Current workaround (not functional, incompatible with functorch)
result = input.clone()
result.narrow(dim, start, length).copy_(src)
```

I'd like to propose adding `torch.narrow_scatter(input, src, dim, start, length) -> Tensor` which scatters values from `src` into a narrow region of `input`:

```python
import torch

# Basic usage
x = torch.zeros(10)
src = torch.ones(5) * 2
result = torch.narrow_scatter(x, src, dim=0, start=2, length=5)
# tensor([0., 0., 2., 2., 2., 2., 2., 0., 0., 0.])

# Autograd support
x = torch.randn(10, 10, requires_grad=True)
src = torch.randn(5, 10, requires_grad=True)
result = torch.narrow_scatter(x, src, 0, 2, 5)
result.sum().backward()  # gradients computed correctly

# torch.compile support
@torch.compile
def fn(x, src):
    return torch.narrow_scatter(x, src, 0, 2, 5)
```

The implementation can delegate to `slice_scatter` since narrow is just slice with step=1:

```cpp
at::Tensor narrow_scatter_symint(..., SymInt start, SymInt length) {
  c10::SymInt end = start + length;
  return at::slice_scatter_symint(self, src, dim, start, end, 1);
}
```

This keeps the implementation minimal and reuses all existing `slice_scatter` optimizations.

[Related PR #175609](https://github.com/pytorch/pytorch/pull/175609/changes)

### Alternatives

Users can work around this with `slice_scatter` directly:
```python
result = torch.slice_scatter(input, src, dim, start, start+length, 1)
```

However, this is less intuitive (requires computing `end = start + length`) and inconsistent with the `narrow_copy` naming convention.

The mutation-based workaround also exists but doesn't work with functorch:
```python
result = input.clone()
result.narrow(dim, start, length).copy_(src)
```

### Additional context

[Related PR #175609](https://github.com/pytorch/pytorch/pull/175609/changes)

cc @pragupta @bdhirsh @ezyang @chauhang @penguinwu @Chillee @samdow @kshitij12345 @bobrenjc93 @aorenste

### PR DESCRIPTION
**Pull request by Sanskaaar1**
Implements torch.narrow_scatter completing the *_copy/*_scatter family.

Implementation:
- Added definition in native_functions.yaml with SymInt support
- C++ implementation in TensorShape.cpp delegates to slice_scatter
- Autograd derivatives in derivatives.yaml
- torch.compile integration:
  - TorchDynamo tracing (trace_rules.py)
  - Inductor lowering (lowering.py)
  - Re-inplacing optimization (reinplace.py)
  - Noop optimization (post_grad.py)
- OpInfo tests with 54 automated test cases
- Gradcheck tests in test_autograd.py

Testing:
Verified with PYTORCH_TEST_WITH_DYNAMO=1, all tests passing.

Fixes #177372

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @Lucaskabela

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -4448,6 +4448,14 @@
   dispatch:
     CPU: narrow_copy_dense_cpu_out
 
+- func: narrow_scatter(Tensor self, Tensor src, int dim, SymInt start, SymInt length) -> Tensor
+  variants: function, method
+  device_check: NoCheck
+  device_guard: False
+  dispatch:
+    CompositeExplicitAutogradNonFunctional: narrow_scatter_symint
```
Comment (isuruf) [inline aten/src/ATen/native/native_functions.yaml]: `CompositeExplicitAutogradNonFunctional` is used for non-aliasing operator decomposing into a aliasing op, but `narrow_scatter` decomposes into `slice_scatter` which is non-aliasing right? See https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native#dispatch

Comment (Sanskaaar1) [reply aten/src/ATen/native/native_functions.yaml]: Updated the dispatch keyword to `CompositeExplicitAutograd`

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by Sanskaaar1** [comment]
I've added documentation for torch.narrow_scatter in the latest push.
This fixes the docs coverage CI failure. Docs build passes with 100% coverage, 0 undocumented APIs.
cc: @jansel  - Could you please review it?

**Comment by jansel** [comment]
@albanD @ezyang what is the process for approving new ops

**Comment by ezyang** [comment]
It used to be @bdhirsh for the `_scatter` ops specifically but we don't currently have anyone with this paged in

**Comment by ezyang** [comment]
@claude I feel there was some sort of historical reason why we didn't originally add this, can you find it? Maybe because narrow is actually a CompositeImplicitAutograd op?

**Comment by zou3519** [comment]
@claude review these changes

**Review by zou3519 — CHANGES_REQUESTED** [review]
narrow is compositeimplicitautograd, we only want view_scatter / view_copy operations for views that are not compositeimplicitautograd. Please let me know if you disagree with my assessment, if not, let's close this PR.

**Comment by Sanskaaar1** [comment]
That makes sense,
since `narrow` is `CompositeImplicitAutograd`, `slice_scatter` already covers this. 
Thanks for the review!
Closing.
